### ISSUE TITLE
Openssl Suddenly Missing - Disables AWS

### ISSUE THREAD

**Issue by brermike** [bwp91/homebridge-govee#405]
### What issue do you have? Please be as thorough and explicit as possible.

I suddenly started receiving the following error in the log for this plugin and am not sure how to resolve.  I was using AWS but this error causes AWS to be disabled.

[12/28/2022, 12:53:57 PM] [Govee] [HTTP] disabling client as Could not find openssl on your system on this path: openssl at C:\Users\Michael\AppData\Roaming\npm\node_modules\homebridge-govee\node_modules\@bwp91\pem\webpack:\pem\lib\openssl.js:257:1

See more log entries below.

I am completely up-to-date on homebridge and the plugin itself.  As far as I can tell, nothing else with my setup has changed.

### Details of your setup.
* Do you use (1) Homebridge UI-X (2) Homebridge CLI or (3) HOOBS? 

1 Homebridge UI-X

* Which version of Homebridge/HOOBS do you have?

v1.6.0

* Which platform do you run Homebridge/HOOBS on (e.g. Raspberry Pi/Windows/HOOBS Box)? Please also mention your version of Node.js/NPM if known.

Windows 10

* Which version of this plugin (homebridge-govee) do you have? Has the issue started since upgrading from a previous version?

homebridge-govee v7.7.1

I don't know when exactly this started happening but sometime in the last month or so.

* Which Govee devices do you have that are causing issues? Please include product models if applicable.

Govee Dreamview T1 
Model #H6199

### Please paste any relevant logs below.

[12/28/2022, 12:53:57 PM] [Govee] [LAN] client enabled and found 0 device(s).
[12/28/2022, 12:53:57 PM] [Govee] [HTTP] retrieved access token from cache.
[12/28/2022, 12:53:57 PM] [Govee] [HTTP] client enabled and found 1 device(s).
[12/28/2022, 12:53:57 PM] [Govee] [HTTP] disabling client as Could not find openssl on your system on this path: openssl at C:\Users\Michael\AppData\Roaming\npm\node_modules\homebridge-govee\node_modules\@bwp91\pem\webpack:\pem\lib\openssl.js:257:1.
[12/28/2022, 12:53:57 PM] [Govee] [AWS] disabling client as requires HTTP client.
[12/28/2022, 12:53:57 PM] [Govee] [API] client enabled and found 1 device(s).
[12/28/2022, 12:53:57 PM] [Govee] [BLE] disabling client as Cannot find package '@abandonware/bluetooth-hci-socket' imported from C:\Users\Michael\AppData\Roaming\npm\node_modules\homebridge-govee\lib\platform.js at new NodeError (node:internal/errors:393:5).

**Comment by micjgeoghegan** [comment]
I ran into this problem the other day. I had to set homebridge to run as an actual admin account, rather than local service or network service in the services console. After that I restarted the server and it picked up OpenSSL. Also need to be on OpenSSL 1.1.1, not the latest version.

**Comment by brermike** [comment]
Thank you!  Installing OpenSSL 1.1.1 fixed all my issues.  I guess it got upgraded at some point.

**Comment by techdudeinga** [comment]
Hi, I am having the same issue, but I use docker, and am not very experienced with this process, and not to good with Linux.  Can anyone assist?  Same issue, Disabling client as Could not find OpenSSL on your system… 

Thanks!

**Comment by bwp91** [comment]
> I ran into this problem the other day. I had to set homebridge to run as an actual admin account, rather than local service or network service in the services console. After that I restarted the server and it picked up OpenSSL. Also need to be on OpenSSL 1.1.1, not the latest version.

Hi @micjgeoghegan your comment has been really useful, is there any chance you could expand a bit on this like a step by step (if you remember!) unfortunately i don’t have a windows setup so i can’t go through this myself but it would be super useful for a lot of other users if you were able to give some kind of ELI5 guide 😅

**Comment by tanjello** [comment]
Since everyone just says installing OpenSSL worked for me but doesn't actually tell you how to do it, here you go.

https://public.cloud.myinfo.gov.sg/docs/OpenSSL_installation_guide.pdf

Once you're done I don't care what the directions say REBOOT YO MACHINE!


After 2 months I finally got it working.


Cliff notes: 

Once the installation completed:
1. Set Environment Variables:
a. Open System Properties > Advanced > Environment Variables
b. Under System variables, click New then input the following as shown in the 
screenshot
Variable name: OPENSSL_CONF
Variable value: Destination location/folder (from #Step2) + \bin\openssl.cfg
(e.g. C:\Program Files\OpenSSL-Win64\bin\openssl.cfg)
P/s: The variable value is the Path to the OpenSSL program stored on your local computer.
It may differ from the screenshot above.
2. Set Path Variable:
a. Open System Properties > Advanced > Environment Variables
b. Under System variables, select Path from the list then click Edit. (You should come 
to a pop-up window with a list of directory paths)
c. Click New and input the directory path to the OpenSSL bin folder. An example is 
shown in the screenshot below.
P/s: Your Path may differ from the screenshot above

**Comment by bwp91** [comment]
thanks for this @tanjello

**Comment by TheAlphaLaw** [comment]
> Since everyone just says installing OpenSSL worked for me but doesn't actually tell you how to do it, here you go.
> 
> https://public.cloud.myinfo.gov.sg/docs/OpenSSL_installation_guide.pdf
> 
> Once you're done I don't care what the directions say REBOOT YO MACHINE!
> 
> After 2 months I finally got it working.
> 
> Cliff notes:
> 
> Once the installation completed:
> 
> 1. Set Environment Variables:
>    a. Open System Properties > Advanced > Environment Variables
>    b. Under System variables, click New then input the following as shown in the
>    screenshot
>    Variable name: OPENSSL_CONF
>    Variable value: Destination location/folder (from #Step2) + \bin\openssl.cfg
>    (e.g. C:\Program Files\OpenSSL-Win64\bin\openssl.cfg)
>    P/s: The variable value is the Path to the OpenSSL program stored on your local computer.
>    It may differ from the screenshot above.
> 2. Set Path Variable:
>    a. Open System Properties > Advanced > Environment Variables
>    b. Under System variables, select Path from the list then click Edit. (You should come
>    to a pop-up window with a list of directory paths)
>    c. Click New and input the directory path to the OpenSSL bin folder. An example is
>    shown in the screenshot below.
>    P/s: Your Path may differ from the screenshot above

_Tried this on two different Windows 11 machines, but Homebridge/Plugin still doesn't see it and disables AWS._

Edit: It works, thank you! Needed to only use 1.1.1

**Comment by nfgrep** [comment]
Just for posterity; I'm trying out a [soon-to-come PR of homebridge for nixos](https://github.com/NixOS/nixpkgs/pull/262427) and ran into this same message:

```
  [2025-04-26, 11:44:11 p.m.] [Govee] [LAN] client enabled and found 0 device(s).
  [2025-04-26, 11:44:12 p.m.] [Govee] [HTTP] client enabled and found 10 device(s).
  [2025-04-26, 11:44:12 p.m.] [Govee] [HTTP] disabling client as Could not find openssl on your system on this path: openssl at /var/lib/homebridge/node_modules/@homebridge-plugins/homebridge-govee/node_modules/pem/webpack:/pem/lib/openssl.js:258:1.
  [2025-04-26, 11:44:12 p.m.] [Govee] See https://github.com/homebridge-plugins/homebridge-govee/wiki/OpenSSL for more info about installing OpenSSL
```

And what ended up fixing it was this:

```nix
# Make sure we've got openssl
environment.systemPackages = with pkgs; [
 openssl
];

# Put openssl on the PATH for homebridge
systemd.services.homebridge.path = [ pkgs.openssl ];
```

(No action needed here on behalf of the govee plugin, this is just incase a nixos user from the future ends up on this issue like I did).

**Comment by yerudako** [comment]
I didn't want to deal with installing 1.1.1 but had it complain "unable to load provider legacy". I copied the legacy.dll from the distribution to the folder it was looking for it in. It worked right after it. For some reason, the flow does not seem to respect whatever I set in OPENSSL_MODULES or the config.
